### PR TITLE
Decimal class serializer

### LINKED ISSUE #50: Support for Decimal type.
The [Decimal](https://docs.python.org/3/library/decimal.html#decimal-objects) type needs to be supported. This should be supported under the `Numeric` group of types. See the "Adding Support For More Types" section of `CONTRIBUTING.md` for more information.

### PR DESCRIPTION
**Pull request by ysmolyar**
**Describe the Changes**
Implemented Decimal class serialization / deserialization.

**Example Code**

```python
import decimal

data = decimal.Decimal(3.14)

serialized = json.dumps(data, cls=JsonSerializer)
# serialized = {'__snappiershot_numeric__': 'decimal', 'value': 3.14}

deserialized = json.loads(serialized, cls=JsonDeserializer)
# deserialized = decimal.Decimal(3.14)

assert deserialized == data
```

**Additional Notes**
Implemented as a custom numeric type.

### CODE REVIEW

File: snappiershot/serializers/json.py [hunk 1]
```diff
@@ -85,6 +86,7 @@ def encode_numeric(value: Number) -> JsonType:
 
         This will do nothing to naturally serializable types (bool, int, float)
           but will perform custom encoding for non-supported types (complex).
+        This will cast Decimal values to float values.
```
Comment (bbonenfant) [inline snappiershot/serializers/json.py]: This conversion is lossy. I think [the `Decimal.from_float` docs](https://docs.python.org/3.9/library/decimal.html#decimal.Decimal.from_float) summarize this pretty well. We should maybe the `Decimal.as_tuple` method which seems to handle many edge cases for use, such as `-0`

File: tests/test_serializers/test_json.py [hunk 2]
```diff
@@ -18,6 +18,7 @@ class TestNumericEncoding:
 
     NUMERIC_DECODING_TEST_CASES = [
         (3 + 4j, CustomEncodedNumericTypes.complex.json_encoding([3, 4])),
+        (Decimal(3.1415), CustomEncodedNumericTypes.decimal.json_encoding(3.1415)),
```
Comment (bbonenfant) [inline tests/test_serializers/test_json.py]: Having `Decimal("3.1415")` (with number quoted) causes `TestNumericEncoding.test_decode_numeric` to fail.

**Comment by bbonenfant** [comment]
Closes #50

**Comment by bbonenfant** [comment]
Additionally, please make destination branch of this PR be `develop`, not `primary`

File: snappiershot/serializers/constants.py [hunk 3]
```diff
@@ -1,10 +1,11 @@
 """ Constant values used by the serializers. """
 import datetime
 from abc import ABC
-from typing import Any, Dict, Iterator, List, NamedTuple, Set, Union
+from decimal import Decimal
+from typing import Any, Dict, Iterator, List, NamedTuple, Set, Tuple, Union
 
 _Primitive = Union[bool, float, int, None, str]
-JsonType = Union[_Primitive, Dict[str, Any], List[Any]]
+JsonType = Union[_Primitive, Dict[str, Any], List[Any], Tuple[Any, ...]]
```
Comment (bbonenfant) [inline snappiershot/serializers/constants.py]: I don't think changing this type signature affecting anything, so I'd revert it

File: snappiershot/serializers/json.py [hunk 4]
```diff
@@ -100,8 +103,11 @@ def encode_numeric(value: Number) -> JsonType:
             # These types are by default supported by the JSONEncoder base class.
             return value
         if isinstance(value, complex):
-            encoded_value = [value.real, value.imag]
+            encoded_value: List[float] = [value.real, value.imag]
             return CustomEncodedNumericTypes.complex.json_encoding(encoded_value)
+        if isinstance(value, Decimal):
+            encode_value: List[Any] = list(value.as_tuple())
```
Comment (bbonenfant) [inline snappiershot/serializers/json.py]: One thing to consider:
I just ran this locally and the "encoded" `Decimal("3.1415")` is `[0, [3, 1, 4, 1, 5], -4]` which isn't very informative... Although this works, I think a nicer format for this data would be converting this named tuple to a dictionary, which is simple to do.

So in the `encode_numeric` method you would have `encode_value = value.as_tuple()._asdict()`, and in the `decode_numeric` method you would have `return Decimal(DecimalTuple(**value))`. This would make the encoding of `Decimal("3.1415")` be `{"sign": 0, "digits": [3, 1, 4, 1, 5], "exponent": -4}` which is more verbose but much easier to understand.

**Review by bbonenfant — APPROVED** [review]
Make sure to use an informative but concise commit message for the squash and merge

File: snappiershot/serializers/json.py (hunk 4 above)
Comment (ysmolyar) [reply snappiershot/serializers/json.py]: Great idea.

File: snappiershot/serializers/constants.py (hunk 3 above)
Comment (ysmolyar) [reply snappiershot/serializers/constants.py]: Forgot I did this! Good catch.
